src/resmom/linux/mom_start.c

```diff
@@ -2097,6 +2097,37 @@ set_job(job *pjob, struct startjob_rtn *sjr)
 					 */
 			sjr->sj_reservation = 0;
 		}
+
+		/*
+		 * Save the ALPS reservation ID so that it can be used to
+		 * delete the reservation in case of race conditions
+		 * where the ID doesn't get recorded in the job structure.
+		 */
+
+		if(pjob->ji_extended.ji_ext.ji_reservation > 0) {
```

Comment:
@npadole20 I'm confused here, If job structure correctly states the reservation id in it then why to we need to store it in another file. I am surely missing something.

Reply:
this is the child process which will become the job. The parent process is getting killed with CTRL-C, hence the so-called "race".

Reply:
Thanks @prakashcv13  for clarifying this. Another question :) can't the child report the reservation ID to parent MOM before spawning reader/writer for the interactive job?

Reply:
There are chances that the parent MoM gets killed before it could read the reservation ID. We need something independent of the parent MoM.

Reply:
But in this case parent MoM is the pbs_mom, why will it get killed? I think we have not even spawned a reader/writer for interactive job so it is certain that user can not kill anything.
I don't understand why it needs to be independent of parent MoM? If we are considering a case that someone/some-script kills the MoM before it could read reservation ID then it can also happen that the child process gets killed before it writes anything to the file.
What do you think?

Reply:
The issue reported and seen to be happening is that the parent MoM is getting killed by the user who presses CTRL-C. Your suggestion is that we send the ALPS reservation ID to the parent MoM instead of writing to a file, but that will not help as there are chances that the parent MoM has been killed by the user even before the child tries to send the ID to her as I have already said.
When you say that there are chances that the child gets killed, that is possible when the user on purpose kills the child process using a kill command on the child process. In that case, PBS should not be responsible for the orphan ALPS reservation.
By issuing a CTRL-C which a user would normally do to kill an interactive job before it starts running will kill the parent MoM as far as I know.
I hope this helps.

Reply:
Well It still needs to be done in a child because we execv to alps client. I think what we can do is:
- let main mom not spawn reader/writer until it gets the reservation information from child
- child should report the reservation id to parent mom instead of writing to file
- make sure the child itself deletes the reservation if it gets SIGTERM or something

what do you think?

Reply:
I think if it is not possible to have the parent mom to manage the reservation ID, what you say seems to be a good solution.

Reply:
The child process which will become the job later, does a fork to create ALPS reservation. Immediately after the ALPS reservation is created and confirmed, the reservation ID is written to the file. When the user does ctrl +C, the job child process is terminated. But I don't understand how it will affect the process writing the reservation ID or how it can get killed.
When user does ctrl +C, the job is deleted by the server and the job data is not available to the mom and that's why it can't cancel the ALPS reservation.

Reply:
@arungrover making the child itself delete the reservation seems like a good solution although I am not sure how exactly to do it yet. I will need to dig into the code and figure it out.

Reply:
@npadole20 Now that I think about it, I think the case I am trying to address would really be a corner case.  I shall leave it upto you to decide whether you want to continue with this fix or fix it in some other way. I'm okay either way..

Reply:
@arungrover and @npadole20, I had approved these changes earlier thinking that the parent mom was getting killed, later on I came to know that pbs_mom is the parent. Understanding the scenario I feel the solution of writing the ID to a file is just narrowing down the window in which the issue can happen. I would suggest that we explore the idea of providing a proper solution, if it turns out that the solution is too complex for the issue at hand, we can go ahead with this.

Reply:
@prakashcv13 I'm fine with that